source/change-streams/tests/README.rst

```diff
@@ -0,0 +1,156 @@
+drivers can use to prove their conformance to the Change Streams Spec.
+
+Several prose tests, which are not easily expressed in YAML, are also presented
+in this file. Those tests will need to be manually implemented by each driver.
+
+Spec Test Format
+================
+
+Each YAML files has the following keys:
+
+- ``database_name``: The default database
+- ``collection_name``: The default collection
+- ``tests``: An array of tests that are to be run independently of each other.
+  Each test will have some of the following fields:
+
+  - ``description``: The name of the test.
+  - ``minServerVersion``: The minimum server version to run this test against. If not present, assume there is no minimum server version.
+  - ``maxServerVersion``: Reserved for later use
+  - ``failPoint``: Reserved for later use
+  - ``target``: The entity to run a changeStream on. Valid values are:
+  
+    - ``collection``: Watch changes on collection ``database_name.collection_name``
+    - ``db``: Watch changes on database ``database_name``
+    - ``client``: Watch changes on entire clusters
+  - ``topology``: An array of server topologies to run the test against
+    Valid topologies are ``single``, ``replicaset``, and ``sharded``.
+  - ``changeStreamPipeline``: An array of additional aggregation pipeline stages to add to the change stream
+  - ``changeStreamOptions``: Additional options to add to the changeStream
+  - ``operations``: Array of documents, each describing an operation. Each document has the following fields:
+    - ``database``: Database to run the operation against
+    - ``collection``: Collection to run the operation against
+    - ``commandName``: Name of the command to run
+    - ``arguments``: Array of arguments for the command.
+
+  - ``expectations``: Optional list of command-started events
+  - ``result``: Document with ONE of the following fields:
+
+    - ``error``: Describes an error received during the test
+    - ``success``: An array of documents expected to be received from the changeStream
+
+Spec Test Match Function
+========================
+
+The definition of MATCH or MATCHES in the Spec Test Runner is as follows:
+
+- MATCH takes two values, ``expected`` and ``actual``
+- Notation is "Assert [actual] MATCHES [expected]
+- Assertion passes if ``expected`` is a subset of ``actual``, with the values ``42`` and ``"42"`` acting as placeholders for "any value"
+
+Pseudocode Implementation of ``actual`` MATCHES ``expected``:
+
+- If ``expected`` is ``"42"`` or ``42``
+
+  - Assert that ``actual`` exists (is not ``null`` or ``undefined``)
+
+- Else
+
+  - Assert that ``actual`` is of the same JSON type as ``expected``
+  - If ``expected`` is a ``number``, ``string``, ``boolean`` or ``null``
+
+    - Assert that ``expected`` equals ``actual``
+
+  - Else if ``expected`` is an ``array``
+
+    - for every ``idx``/``value`` in ``expected``
+
+      - Assert that ``value`` MATCHES ``actual[idx]``
+
+  - Else
```

Comment:
Considering the earlier conditions, this assumes that all possible types are limited to:

 * number
 * string
 * boolean
 * null
 * array
 * object (this else case)

Would it be safer to explicitly check for an object here and either omit a final `else` or use it to report some error for an unsupported type?

Reply:
Those are all of the valid JSON types. We could also rework it to be:

```
If expected is a JSON object:
  ...
Else if expected is a JSON array:
  ...
Else:
  Assert that expected equals actual
```

Reply:
Noted. I see you reworked it to make the non-object/array case the `else` condition with a basic equality check. I think that works nicely.